Remove `constructor` from `HighlightRegistry.idl`

Ahmad-S792 opened the pull request:
#### 10396bf507cc815f725b7c17bad57e7b93b84e2b
<pre>
Remove `constructor` from `HighlightRegistry.idl`
<a href="https://bugs.webkit.org/show_bug.cgi?id=271360">https://bugs.webkit.org/show_bug.cgi?id=271360</a>
<a href="https://rdar.apple.com/problem/125529396">rdar://problem/125529396</a>

Reviewed by Brent Fulgham.

This patch aligns WebKit with Gecko / Firefox, Blink / Chromium and
Web Specification [1]:

[1] <a href="https://drafts.csswg.org/css-highlight-api/#highlightregistry">https://drafts.csswg.org/css-highlight-api/#highlightregistry</a>

This removes `constructor` since it is not in web specification.

Additionally, this patch addresses two issues with the HighlightRegistry
tampered prototype test:

1. Removes reliance on the non-standard HighlightRegistry constructor and
   instead uses CSS.highlights to access the standard HighlightRegistry
   instance.

2. Properly restores Map.prototype to its original state after the test
   completes. Since Map.prototype is frozen during the test, restoration is
   done by creating a new prototype object with the saved descriptors and
   replacing Map&apos;s prototype, rather than attempting to modify the frozen
   object directly.

NOTE: Both of above changes were suggested by Alexey Shvaika (original
test author).

* LayoutTests/highlight/highlight-interfaces-expected.txt:
* LayoutTests/highlight/highlight-interfaces.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/idlharness.window-expected.txt:
* Source/WebCore/Modules/highlight/HighlightRegistry.idl:
* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/HighlightRegistry-maplike-tampered-Map-prototype.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/HighlightRegistry-maplike-expected.txt:

Canonical link: <a href="https://commits.webkit.org/303858@main">https://commits.webkit.org/303858@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/08fdc388c6c936fdc14fba7c5e9db1d73b50671f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133119 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5620 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44229 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140664 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85158 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/fed90578-7d87-4d64-ae6b-df1491c79d05) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134989 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6118 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5483 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101807 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69185 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/31703ecf-2620-4236-8236-c6c4e6b623be) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136066 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4317 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119300 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82604 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7c2045d3-37ad-4859-bac9-0084299864eb) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4201 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1786 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113276 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37416 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143311 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5290 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37995 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110186 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5372 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4541 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110366 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28131 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4083 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115557 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58952 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5345 "Built successfully") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33911 "Found 1 new test failure: fast/mediastream/granted-denied-request-management1.html (failure)") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5188 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68797 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5434 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5301 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa reviewed (commented):
Hm... this does seem like a compatibility concern. The code which was written for Safari will start throwing. Not sure if it's worth the risk.

brentfulgham commented on LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/HighlightRegistry-maplike-tampered-Map-prototype-expected.txt:
```diff
@@ -1,3 +1,3 @@
 
-PASS HighlightRegistry is a maplike interface that works as expected even if Map.prototype is tampered.
+FAIL HighlightRegistry is a maplike interface that works as expected even if Map.prototype is tampered. Illegal constructor
```
Why is this test failing? Shouldn't removing the `constructor` cause it to pass?

Ahmad-S792 replied:
It is failing for all browsers and was upstreamed by WebKit. It relies on `constructor`, which web specification and other browsers don't add.

annevk replied:
We should really fix the test here.

Ahmad-S792 replied:
@annevk - yes, I got some insight from Alexey, so will try to look into it tomorrow or day after. :-)

brentfulgham reviewed (approved):
Looks great! Thank you for fixing this old bug! r=me

Commit:
Remove `constructor` from `HighlightRegistry.idl`
https://bugs.webkit.org/show_bug.cgi?id=271360
rdar://problem/125529396

Reviewed by Brent Fulgham.

This patch aligns WebKit with Gecko / Firefox, Blink / Chromium and
Web Specification [1]:

[1] https://drafts.csswg.org/css-highlight-api/#highlightregistry

This removes `constructor` since it is not in web specification.

Additionally, this patch addresses two issues with the HighlightRegistry
tampered prototype test:

1. Removes reliance on the non-standard HighlightRegistry constructor and
   instead uses CSS.highlights to access the standard HighlightRegistry
   instance.

2. Properly restores Map.prototype to its original state after the test
   completes. Since Map.prototype is frozen during the test, restoration is
   done by creating a new prototype object with the saved descriptors and
   replacing Map's prototype, rather than attempting to modify the frozen
   object directly.

NOTE: Both of above changes were suggested by Alexey Shvaika (original
test author).

* LayoutTests/highlight/highlight-interfaces-expected.txt:
* LayoutTests/highlight/highlight-interfaces.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/idlharness.window-expected.txt:
* Source/WebCore/Modules/highlight/HighlightRegistry.idl:
* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/HighlightRegistry-maplike-tampered-Map-prototype.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/HighlightRegistry-maplike-expected.txt:

Canonical link: https://commits.webkit.org/303858@main
